source/opt/combine_access_chains.cpp

```diff
@@ -0,0 +1,253 @@
+
+  Instruction* base_ptr = def_use_mgr->GetDef(inst->GetSingleWordInOperand(0));
+  const analysis::Type* type = type_mgr->GetType(base_ptr->type_id());
+  assert(type->AsPointer());
+  type = type->AsPointer()->pointee_type();
+  std::vector<uint32_t> element_indices;
+  uint32_t starting_index = 1;
+  if (inst->opcode() == SpvOpPtrAccessChain) {
+    // Skip the first index of OpPtrAccessChain as it does not affect type
+    // resolution.
+    starting_index = 2;
+  }
+  for (uint32_t i = starting_index; i < inst->NumInOperands(); ++i) {
+    Instruction* index_inst =
+        def_use_mgr->GetDef(inst->GetSingleWordInOperand(i));
+    const analysis::Constant* index_constant =
+        context()->get_constant_mgr()->GetConstantFromInst(index_inst);
+    if (index_constant) {
+      uint32_t index_value = GetConstantValue(index_constant);
+      element_indices.push_back(index_value);
+    } else {
+      // This index must not matter to resolve the type in valid SPIR-V.
+      element_indices.push_back(0);
+    }
+  }
+  type = type_mgr->GetMemberType(type, element_indices);
+  return type;
+}
+
+bool CombineAccessChains::CombineIndices(Instruction* ptr_input,
+                                         Instruction* inst,
+                                         std::vector<Operand>* new_operands) {
+  analysis::DefUseManager* def_use_mgr = context()->get_def_use_mgr();
+  analysis::ConstantManager* constant_mgr = context()->get_constant_mgr();
+
+  Instruction* last_index_inst = def_use_mgr->GetDef(
+      ptr_input->GetSingleWordInOperand(ptr_input->NumInOperands() - 1));
+  const analysis::Constant* last_index_constant =
+      constant_mgr->GetConstantFromInst(last_index_inst);
+
+  Instruction* element_inst =
+      def_use_mgr->GetDef(inst->GetSingleWordInOperand(1));
+  const analysis::Constant* element_constant =
+      constant_mgr->GetConstantFromInst(element_inst);
+
+  // Combine the last index of the AccessChain (|ptr_inst|) with the element
+  // operand of the PtrAccessChain (|inst|).
+  const bool combining_element_operands =
+      inst->opcode() == SpvOpPtrAccessChain &&
+      ptr_input->opcode() == SpvOpPtrAccessChain &&
+      ptr_input->NumInOperands() == 2;
+  uint32_t new_value_id = 0;
+  const analysis::Type* type = GetIndexedType(ptr_input);
+  if (last_index_constant && element_constant) {
+    // Combine the constants.
+    uint32_t new_value = GetConstantValue(last_index_constant) +
+                         GetConstantValue(element_constant);
+    const analysis::Constant* new_value_constant =
+        constant_mgr->GetConstant(last_index_constant->type(), {new_value});
+    Instruction* new_value_inst =
+        constant_mgr->GetDefiningInstruction(new_value_constant);
+    new_value_id = new_value_inst->result_id();
+  } else if (!type->AsStruct() || combining_element_operands) {
+    // TODO(alan-baker): handle this unlikely case.
+    if (last_index_inst->type_id() != element_inst->type_id()) return false;
```

Comment:
Actually, the signedness semantics depends on the consuming SPIR-V instruction, not on the type of the value.
(It's that subtle rule...)

So I think you don't have to bail here.